src/devices/machine/ibmps2.cpp

```diff
@@ -0,0 +1,1113 @@
+	m_mcabus->drq_callback<0>().set(m_dmac, FUNC(ibm72x7377_device::dreq_w<0>));
+	m_mcabus->drq_callback<1>().set(m_dmac, FUNC(ibm72x7377_device::dreq_w<1>));
+	m_mcabus->drq_callback<2>().set(m_dmac, FUNC(ibm72x7377_device::dreq_w<2>));
+	m_mcabus->drq_callback<3>().set(m_dmac, FUNC(ibm72x7377_device::dreq_w<3>));
+	m_mcabus->drq_callback<5>().set(m_dmac, FUNC(ibm72x7377_device::dreq_w<5>));
+	m_mcabus->drq_callback<6>().set(m_dmac, FUNC(ibm72x7377_device::dreq_w<6>));
+	m_mcabus->drq_callback<7>().set(m_dmac, FUNC(ibm72x7377_device::dreq_w<7>));
+	m_mcabus->iochck_callback().set(FUNC(ps2_mb_device::iochck_w));
+	m_mcabus->cs_feedback_callback().set(FUNC(ps2_mb_device::cd_sfdbk_w));
+}
+
+
+void ps2_mb_device::device_add_mconfig(machine_config &config)
+{
+	if(m_sram_size != 0) NVRAM(config, m_nvram);
+
+	// A lot of this stuff is integrated into the 72X8299 I/O controller.
+	PS2_PIT(config, m_pit);
+	m_pit->set_clk<0>(XTAL_IO / 12); /* heartbeat IRQ */
+	m_pit->out_handler<0>().set(FUNC(ps2_mb_device::irq0_w));
+	m_pit->set_clk<1>(XTAL_IO / 12); /* dram refresh */
+	m_pit->set_clk<2>(XTAL_IO / 12); /* pio port c pin 4, and speaker polling enough */
+	m_pit->out_handler<2>().set(FUNC(ps2_mb_device::pit8254_out2_changed));
+	// CH 3 behavior: IRQ0 asserted enables the gate.
+	// CLK IN 3 is tied to ~(CLK OUT 0).
+	// If CLK OUT 3 is high, the NMI is asserted.
+	// Effectively, it asserts NMI when IRQ 0 is active for more than one CH 0 tick.
+	m_pit->ch3_out_handler().set(FUNC(ps2_mb_device::watchdog_w));
+
+	IBM72X7377(config, m_dmac, 14.318181_MHz_XTAL / 3);
+
+	PIC8259(config, m_pic8259_master);
+	m_pic8259_master->out_int_callback().set_inputline(m_maincpu, 0);
+	m_pic8259_master->in_sp_callback().set_constant(1);
+	m_pic8259_master->read_slave_ack_callback().set(FUNC(ps2_mb_device::get_slave_ack));
+
+	PIC8259(config, m_pic8259_slave);
+	m_pic8259_slave->out_int_callback().set(m_pic8259_master, FUNC(pic8259_device::ir2_w));
+	m_pic8259_slave->in_sp_callback().set_constant(0);
+
+	MC146818(config, m_mc146818, 32.768_kHz_XTAL);
+	m_mc146818->irq().set(m_pic8259_slave, FUNC(pic8259_device::ir0_w));
+	m_mc146818->set_century_index(0x32);
+
+	/* sound hardware */
+	SPEAKER(config, "mono").front_center();
+	SPEAKER_SOUND(config, m_speaker).add_route(ALL_OUTPUTS, "mono", 0.50);
+
+	PC_KBDC(config, m_ps2_con, pc_ps2_keyboards, STR_KBD_MICROSOFT_NATURAL);
+	m_ps2_con->out_clock_cb().set(m_keybc, FUNC(ps2_keyboard_controller_device::kbd_clk_w));
+	m_ps2_con->out_data_cb().set(m_keybc, FUNC(ps2_keyboard_controller_device::kbd_data_w));
+
+	PC_KBDC(config, m_aux_con, ps2_mice, STR_HLE_PS2_MOUSE);
+	m_aux_con->out_clock_cb().set(m_keybc, FUNC(ps2_keyboard_controller_device::aux_clk_w));
+	m_aux_con->out_data_cb().set(m_keybc, FUNC(ps2_keyboard_controller_device::aux_data_w));
+
+	PS2_KEYBOARD_CONTROLLER(config, m_keybc, 12_MHz_XTAL);
+	m_keybc->set_default_bios_tag("ibm");
+	m_keybc->hot_res().set_inputline(":maincpu", INPUT_LINE_RESET);
```

Comment:
Once again, assumptions like this are a bad idea – either use a trampoline devcb, or add a trampoline accessor for this callback.